halo2_gadgets/src/poly_eval.rs

```diff
@@ -0,0 +1,484 @@
+    /// Variable representing a coefficient.
+    /// Certainly it could be from another Field/Group,
+    /// orther than current one
+    type Coeff;
+
+    /// Field element
+    type Elem;
+
+    /// Load Element to be used as eval point
+    fn load_element(&self, layouter: impl Layouter<F>, a: Value<F>) -> Result<Self::Elem, Error>;
+
+    /// Evaluation
+    fn eval(
+        &self,
+        layouter: impl Layouter<F>,
+        coeffs: Vec<Self::Coeff>,
+        elem: Self::Elem,
+    ) -> Result<Self::Elem, Error>;
+}
+
+/// Chip for poly eval
+#[derive(Clone, Debug)]
+pub struct PolyEvalConfig {
+    /// column layput : | eval point | coeff | accumulative result |
+    advice: [Column<Advice>; EVALADVCOL],
+
+    /// A selector controls gate
+    e_sel: Selector,
+
+    /// A selector indicating a certain cell is zero
+    zero_sel: Selector,
+}
+
+/// Poly eval chip
+#[derive(Clone, Debug)]
+pub struct PolyEvalChip<F: Field> {
+    config: PolyEvalConfig,
+    _marker: PhantomData<F>,
+}
+
+impl<F: Field> PolyEvalChip<F> {
+    /// Construct a poly eval chip, return a structured config
+    pub fn construct(config: <Self as Chip<F>>::Config) -> Self {
+        Self {
+            config,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Define the gates that constraint system needs
+    /// Polynomial Evaluation -- Horner's rule
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        advice: [Column<Advice>; EVALADVCOL],
+    ) -> <Self as Chip<F>>::Config {
+        for col in advice.iter() {
+            meta.enable_equality(*col);
+        }
+        // All constraints enforced at once
+        let e_sel = meta.selector();
+        let zero_sel = meta.selector();
+
+        meta.create_gate("zero test", |meta| {
+            // Cells participating in zero test
+            let init_acc = meta.query_advice(advice[ACCINDEX], Rotation::cur());
+            let zero_s = meta.query_selector(zero_sel);
+
+            // Simply making a cell to zero is not sufficient
+            // We have to enforce the relationship between two rows:
+            // That is, the next line acc calucation must be dependent on an inital `acc` with zero value
+            let init_acc_clone = meta.query_advice(advice[ACCINDEX], Rotation::cur());
+            let next_acc = meta.query_advice(advice[ACCINDEX], Rotation::next());
+            let next_coeff = meta.query_advice(advice[COEFFINDEX], Rotation::next());
+            let next_x = meta.query_advice(advice[XINDEX], Rotation::next());
+            let zero_s_clone = meta.query_selector(zero_sel);
+
+            // First, to constrain init acc is zero;
+            // Second, to make sure init acc is guaranteed to be used in following calculation
+            // Namely acc(i+1) = coeff(i) + x * acc(i) recursion formula
+            vec![
+                zero_s * init_acc,
+                zero_s_clone * (next_coeff + next_x * init_acc_clone - next_acc),
```

Comment:
I see. Agreed. Only one thing, `e_sel` shall still start at row 1.

Let's draw it in pic:
| - | - | x |  -> row 0  zero_sel: 1  e_sel: 0
| x | x | x | -> row 1   zero_sel: 0  e_sel: 1
Recall that we have three columns. `x` stands for the constraint covers, and it looks like __| . If we let `e_sel` enabled at row 0 it will hit `cell not assigned` error on circuit. But yes, probably one "__|" alike constraint is sufficient